### PR TITLE
feat(flags): support early_exit in local evaluation

### PR DESCRIPTION
**Pull request by gustavohstrassburger**
## :bulb: Motivation and Context

PostHog added an "early exit" option to feature flags in the server-side (Rust) evaluation engine, and this has already been ported to posthog-node and posthog-python. This PR ports the same behavior to the .NET SDK's local feature flag evaluation.

When a flag's `filters.early_exit` is `true`, local condition evaluation must stop and return a definitive disabled result as soon as a condition group's property filters match (or the group has no property filters) but the rollout percentage excludes the user — instead of falling through to evaluate later condition groups.

The three per-group outcomes are now modeled explicitly, mirroring the Rust engine:

- **MATCH** — property filters matched (or none) AND rollout included the user → flag matches (existing behavior).
- **NO_MATCH** — a property filter did not match → always continue to the next group, even when `early_exit` is enabled.
- **OUT_OF_ROLLOUT_BOUND** — property filters matched (or none) but the rollout excluded the user → with `early_exit` enabled, return a definitive disabled result immediately.

When `early_exit` is absent or `false`, existing behavior is preserved exactly.

### Changes

- Added the `early_exit` field (`EarlyExit`, JSON `early_exit`, default `false`) to `FeatureFlagFilters`, including equality/hash-code.
- Converted the single-group matcher `IsConditionMatch` (bool) into `EvaluateConditionMatch` returning a tri-state `ConditionMatchResult` enum, distinguishing rollout-exclusion (`hash > rollout/100`) from property-mismatch.
- Short-circuit to `false` in the condition-group loop when `early_exit` is on and the outcome is `OutOfRolloutBound`.

## :green_heart: How did you test it?

Added unit tests in `TheEarlyExitBehavior` covering:

1. `early_exit` enabled → `false` without evaluating a later matching group (first group: matching props + rollout 0; second group: matching props + rollout 100).
2. `early_exit` unset (absent in JSON) → falls through to the matching second group (`true`) — regression.
3. `early_exit` explicitly `false` → falls through (`true`).
4. Property-filter failure (not rollout) does NOT early-exit even when `early_exit` is enabled → falls through (`true`).

All 889 existing unit tests still pass (`dotnet test`, net8.0). Release build is clean across all target frameworks with warnings-as-errors, and `dotnet format --verify-no-changes` reports no formatting issues.

## :pencil: Checklist

- [x] I reviewed the submitted code.
- [x] I added tests to verify the changes.
- [ ] I updated the docs if needed.
- [x] No breaking change or entry added to the changelog.

### If releasing new changes

- [x] Ran `pnpm changeset` to generate a changeset file

---
*Created with [PostHog Code](https://posthog.com/code?ref=pr)*

### CODE REVIEW

**Comment by gustavohstrassburger** [comment]
@copilot fix the errors in https://github.com/PostHog/posthog-dotnet/actions/runs/27026012051/job/79766113172?pr=221

Error: /home/runner/work/posthog-dotnet/posthog-dotnet/tests/UnitTests/Features/LocalEvaluatorTests.cs(2708,23): error WHITESPACE: Fix whitespace formatting. Delete 1 characters. [/home/runner/work/posthog-dotnet/posthog-dotnet/tests/UnitTests/UnitTests.csproj]
Error: /home/runner/work/posthog-dotnet/posthog-dotnet/tests/UnitTests/Features/LocalEvaluatorTests.cs(2708,46): error WHITESPACE: Fix whitespace formatting. Delete 1 characters. [/home/runner/work/posthog-dotnet/posthog-dotnet/tests/UnitTests/UnitTests.csproj]
Error: /home/runner/work/posthog-dotnet/posthog-dotnet/tests/UnitTests/Features/LocalEvaluatorTests.cs(2709,46): error WHITESPACE: Fix whitespace formatting. Delete 1 characters. [/home/runner/work/posthog-dotnet/posthog-dotnet/tests/UnitTests/UnitTests.csproj]
Error: /home/runner/work/posthog-dotnet/posthog-dotnet/tests/UnitTests/Features/LocalEvaluatorTests.cs(2710,23): error WHITESPACE: Fix whitespace formatting. Delete 1 characters. [/home/runner/work/posthog-dotnet/posthog-dotnet/tests/UnitTests/UnitTests.csproj]
Error: /home/runner/work/posthog-dotnet/posthog-dotnet/tests/UnitTests/Features/LocalEvaluatorTests.cs(2710,46): error WHITESPACE: Fix whitespace formatting. Delete 1 characters. [/home/runner/work/posthog-dotnet/posthog-dotnet/tests/UnitTests/UnitTests.csproj]

File: src/PostHog/Features/LocalEvaluator.cs [hunk 1]
```diff
@@ -298,9 +299,19 @@ StringOrValue<bool> MatchFeatureFlagProperties(
...
-                // the matching variant
-                if (!IsConditionMatch(flag, effectiveBucketingId, condition, effectiveProperties, evaluationCache, groups))
+                var conditionResult = EvaluateConditionMatch(flag, effectiveBucketingId, condition, effectiveProperties, evaluationCache, groups);
+
+                // When early_exit is enabled and the condition's property filters matched (or
+                // there were none) but the rollout excluded the user, stop evaluating later
+                // condition groups and return a definitive disabled result. Mirrors the
+                // server-side Rust evaluation engine. A pure property-filter mismatch always
+                // falls through to the next condition group, even when early_exit is enabled.
+                if (earlyExit && conditionResult == ConditionMatchResult.OutOfRolloutBound)
```
Comment (dustinbyrne) [inline src/PostHog/Features/LocalEvaluator.cs]: Blocking correctness issue: this can mask a prior inconclusive condition.

If an earlier condition caught `InconclusiveMatchException`, `isInconclusive` is set, but this early-exit branch returns `false` before the post-loop fallback at lines 342-344 can throw. That makes local evaluation return a definitive disabled result when it should fall back to server evaluation.

Repro shape:
1. earlier condition is inconclusive locally, e.g. missing required property
2. later condition evaluates to `OutOfRolloutBound`
3. `early_exit=true`

Current result: `false`
Expected: throw `InconclusiveMatchException` / server fallback

Please check `isInconclusive` before returning `false` here; if it’s true, preserve fallback semantics instead of short-circuiting.

**Review by dustinbyrne — COMMENTED** [review]
Validated the early_exit edge case locally; leaving one blocking correctness note inline.

File: tests/UnitTests/Features/LocalEvaluatorTests.cs [hunk 2]
```diff
@@ -2600,4 +2600,166 @@ public void DoesNotMatchTypeMismatches(object expected, object actual, bool shou
...
+            [
+                new LocalFeatureFlag
+                {
+                    Id = 42,
+                    TeamId = 23,
+                    Name = "early-exit-feature-flag",
+                    Key = "early-exit",
+                    Filters = new FeatureFlagFilters
+                    {
+                        EarlyExit = earlyExit,
```
Comment (haacked) [inline tests/UnitTests/Features/LocalEvaluatorTests.cs:2628]: `suggestion`: No test ever deserializes `early_exit: true` from JSON. Every positive early-exit case sets `EarlyExit = earlyExit` through this object initializer, which never touches the [`[JsonPropertyName("early_exit")]`](https://github.com/posthog/posthog-dotnet/blob/b5b0488675fec02ca7eb6dcaf3a89dacc4d6da51/src/PostHog/Api/LocalEvaluationApiResult.cs#L154) binding. `CreateFlagsWithoutEarlyExitField` only proves the absent field defaults to `false`. So a typo in that property name would pass every test here while real payloads carrying `"early_exit": true` would deserialize to `false` and quietly disable the feature in production. Worth one JSON round-trip that injects `"early_exit": true` and asserts the out-of-rollout group early-exits.

**Review by dustinbyrne — COMMENTED** [review]
Take a look at the following test, it's still a failing condition. Once the early exit / out of rollout band condition is reached, later groups shouldn't be considered. But because an earlier condition was inconclusive, local evaluation also can’t safely return false, so it should preserve the inconclusive/server-fallback path.

```csharp
[Fact]
public void ThrowsInconclusiveWhenEarlyExitOutOfRolloutHasLaterMatchingGroup()
{
    var flags = new LocalEvaluationApiResult
    {
        Flags =
        [
            new LocalFeatureFlag
            {
                Id = 42,
                TeamId = 23,
                Name = "early-exit-feature-flag",
                Key = "early-exit",
                Filters = new FeatureFlagFilters
                {
                    EarlyExit = true,
                    Groups =
                    [
                        new FeatureFlagGroup
                        {
                            Properties =
                            [
                                new PropertyFilter
                                {
                                    Type = FilterType.Person,
                                    Key = "missing_property",
                                    Value = new PropertyFilterValue("some_value"),
                                    Operator = ComparisonOperator.Exact
                                }
                            ],
                            RolloutPercentage = 100
                        },
                        new FeatureFlagGroup
                        {
                            Properties =
                            [
                                new PropertyFilter
                                {
                                    Type = FilterType.Person,
                                    Key = "email",
                                    Value = new PropertyFilterValue("tyrion@example.com"),
                                    Operator = ComparisonOperator.Exact
                                }
                            ],
                            RolloutPercentage = 0
                        },
                        new FeatureFlagGroup
                        {
                            Properties =
                            [
                                new PropertyFilter
                                {
                                    Type = FilterType.Person,
                                    Key = "email",
                                    Value = new PropertyFilterValue("tyrion@example.com"),
                                    Operator = ComparisonOperator.Exact
                                }
                            ],
                            RolloutPercentage = 100
                        }
                    ]
                }
            }
        ],
        GroupTypeMapping = new Dictionary<string, string>()
    };

    var localEvaluator = new LocalEvaluator(flags);

    Assert.Throws<InconclusiveMatchException>(() => localEvaluator.EvaluateFeatureFlag(
        key: "early-exit",
        distinctId: "1234",
        personProperties: new Dictionary<string, object?> { ["email"] = "tyrion@example.com" }));
}
```

**Review by haacked — COMMENTED** [review]
Agree with @dustinbyrne's suggestions. Also, one test suggestion.

File: src/PostHog/Features/LocalEvaluator.cs [hunk 3]
```diff
@@ -338,7 +338,14 @@ StringOrValue<bool> MatchFeatureFlagProperties(
             catch (InconclusiveMatchException)
             {
                 // Evaluation error (bad regex, invalid date, missing property, etc.)
-                // Track that we had an inconclusive match, but try other conditions
+                // If early_exit is enabled and a prior condition was already inconclusive,
+                // we cannot safely skip to later groups — re-throw so the caller falls back
+                // to the server, which mirrors what would have happened had the rollout
+                // boundary short-circuited first.
+                if (earlyExit && isInconclusive)
```
Comment (dustinbyrne) [inline src/PostHog/Features/LocalEvaluator.cs]: `earlyExit` reflects the flag level setting, yes? It's not indicating that for this specific evaluation, we've missed the rollout band, and we should _now_ exit early.

That makes this too broad, and the following test case would now fail:

Given a flag with `EarlyExit: true` and the following condition groups:
1. inconclusive group with rollout 100
2. another inconclusive group with rollout 100
3. matching group with rollout 100

This case should still return `true`, but it'll now throw `InconclusiveMatchException` before reaching the last matching group

Comment (gustavohstrassburger) [reply src/PostHog/Features/LocalEvaluator.cs]: yeah, you're right, I'll fix this, thanks dustin.

**Review by dustinbyrne — APPROVED** [review]
nice!
